### PR TITLE
monitor-fds: Fix on macOS.

### LINKED ISSUE #1871: Daemon doesn't notice client disconnects on macOS
Not sure how to reproduce this exactly, but I've seen several people hit a case on macOS with 1.12pre5788_e3013543 where they cancelled all running builds but build locks were still held and needed to restart the daemon.

### PR DESCRIPTION
**Pull request by shlevy**
Fixes #1871.

### CODE REVIEW

**Comment by shlevy** [comment]
Testing this on my box now, but based on the simplified tests I fully expect it to work.

File: src/libutil/monitor-fd.hh [hunk 1]
```diff
@@ -24,7 +24,12 @@ public:
             /* Wait indefinitely until a POLLHUP occurs. */
             struct pollfd fds[1];
             fds[0].fd = fd;
-            fds[0].events = 0;
+            /* This shouldn't be necessary, but macOS doesn't seem to
+               like a zeroed out events field.
+               See https://openradar.appspot.com/radar?id=5047895313088512
```
Comment (copumpkin) [inline src/libutil/monitor-fd.hh]: Lovely, thanks!

**Comment by shlevy** [comment]
Ugh, hitting the assertion on the next line, digging in...

**Comment by shlevy** [comment]
poll is timing out even with a `-1` timeout. Going to try select...

**Comment by shlevy** [comment]
Seems related... https://github.com/curl/curl/issues/1057

**Comment by shlevy** [comment]
OK, this is tested and working. Ugly, but so is the problem domain.

File: src/libutil/monitor-fd.hh [hunk 2]
```diff
@@ -21,13 +21,23 @@ public:
...
+              fds[0].fd = fd;
+              /* This shouldn't be necessary, but macOS doesn't seem to
+                 like a zeroed out events field.
+                 See https://openradar.appspot.com/radar?id=5047895313088512
+                 and https://bugreport.apple.com/web/?problemID=37537852.
+              */
+              fds[0].events = POLLHUP;
+              auto count = poll(fds, 1, -1);
+              if (count == -1) abort(); // can't happen
+              if (count == 0) continue; // shouldn't happen, but macOS is terrbile
```
Comment (shlevy) [inline src/libutil/monitor-fd.hh]: We may want to insert a delay here if this turns out to cause lots of spinning.

**Comment by shlevy** [comment]
Going to file another mac bug and will update the comment with the links, but we can merge this whenever and I'll add it when it's available.

**Pull request merged**